### ISSUE TITLE
Loss not changing in TFLearn, but model running in Keras (backed by TF) trains well...

### ISSUE THREAD

**Issue by reesepathak** [tflearn/tflearn#253]
## The problem

I built a 3-layer CNN in TF Learn and was unable to learn anything. It gets stuck at a loss of 0.69315 even after 17 iterations. Comparably, in Keras, if I train for 17 iterations, the loss continuously decreases to ~0.07. I think something might be wrong with the way the data is input? I need to reshape it into the form of [#samples, 4,500, 1] for TFLearn, but that's the only difference, for the rest, the models are exactly the same. Assume that the simulation data is correct (as it has nothing to do w/TF and works with Keras). 

Has anyone had any luck? I tried debugging by looking at the TensorBoard output, and it strangely immediately jumps to 100% accuracy? I'm very confused. It isn not as if my target vector is all the same class....

```
# coding: utf-8

# ### Testing out new extension of Google's TensorFlow : TFLearn

# In[1]:

from __future__ import division, print_function, absolute_import
import numpy as np
import seaborn as sns
import tflearn as tfl
import tensorflow as tf
from sklearn.cross_validation import train_test_split
from dragonn.simulations import *
# TF/TF-learn specific prelims
from tflearn.layers.core import input_data, dropout, fully_connected
from tflearn.layers.conv import conv_2d, max_pool_2d
from tflearn.layers.normalization import local_response_normalization
from tflearn.layers.estimator import regression


# ### DATA: Using (legacy) simulation code

# In[33]:

# CTCF Known 2, 1000 base pair sequence, 5000 examples + and -, 40% GC
X, y = simulate_single_motif_detection('CTCF_known2', 500, 5000, 5000, 0.4)


# In[34]:

X = one_hot_encode(X)


# In[35]:

random_order = np.arange(len(X))
np.random.shuffle(random_order)
X, y = X[random_order], y[random_order]


# In[36]:

# Create split datasets
test_fraction = 0.2
X_train, X_test, y_train, y_test = train_test_split(X, y, test_size=test_fraction)
X_train, X_valid, y_train, y_valid = train_test_split(X_train, y_train, test_size=test_fraction)


# In[37]:

# Randomize
X_train = np.concatenate((X_train, reverse_complement(X_train)))
y_train = np.concatenate((y_train, y_train))
X_test = np.concatenate((X_test, reverse_complement(X_test)))
y_test = np.concatenate((y_test, y_test))
X_valid= np.concatenate((X_valid, reverse_complement(X_valid)))
y_valid = np.concatenate((y_valid, y_valid))

X_train = X_train.reshape((12800, 4, 500, 1))
X_valid = X_valid.reshape((3200, 4, 500, 1))
X_test = X_test.reshape((4000, 4, 500, 1))


# In[59]:

graph_to_use = tf.Graph()
with graph_to_use.as_default():
    network = input_data(shape=(None, 4, 500, 1))
    network = conv_2d(network, 45, (4, 4), activation='relu', name='First_Convolutional_Layer')
    network = dropout(network, 0.8)
    network = conv_2d(network, 50, (1, 10), activation='relu', name='Second_Convolutional_Layer')
    network = dropout(network, 0.8)
    network = conv_2d(network, 50, (1, 8), activation='relu', name='Third_Convolutional_Layer')
    network = dropout(network, 0.8)
    network = max_pool_2d(network, (1, 25), name='Max_Pool_Layer')
    network = fully_connected(network, 1, activation='sigmoid', name='Fully_connectected_layer')
    network = regression(network, optimizer='adam', loss='binary_crossentropy')
    model = tfl.DNN(network, tensorboard_verbose=0)
    model.fit(X_train, y_train, n_epoch=2, batch_size=128, validation_set=(X_valid, y_valid), show_metric=True,
             run_id='2.0-RP-ReplicatewithKeras-run0002')



```

**Comment by aymericdamien** [comment]
TFLearn and Keras has different default parameters, that could explain your case (such as gradient clipping, weights initialization, ...) You may try again and make sure that these parameters are actually the same too.
Also, can you try to use 'categorical_crossentropy' instead? 'accuracy' metric is meant to works for this objective function. There is indeed a bug when using 'binary_crossentropy'.

**Comment by reesepathak** [comment]
I used categorical cross entropy and now my loss is just 0 at the beginning, which makes no sense. The same exact dataset (just in a different tensor shape) provides a regular loss/accuracy vs epoch plot that you'd expect. Even in the default params were different, surely you wouldn't expect this to automatically get 100% correct right off the bat?

**Comment by aymericdamien** [comment]
Yes, 100% is actually due to the fact that 'accuracy' only calculate the accuracy for binary vector labels. So in your case, it will predict all prediction as '0' and accuracy get the argmax of the prediction, which is also '0', thus 100%.
If you get your loss falling down to 0 very quickly, it may means that your model is diverging.

**Comment by reesepathak** [comment]
My targets are binary vectors though?

**Comment by aymericdamien** [comment]
I mean try like that: `network = fully_connected(network, 2, ...)` if you have 2 classes.

**Comment by reesepathak** [comment]
To clarify, the training examples are pairs of data points (X, y), where X is the input, and is a matrix 4 X 1000, and the labels are scalar, and are binary categorical labels in the space {0, 1}.

**Comment by reesepathak** [comment]
wait...I shouldn't need to do that. my y labels are either 0 or 1. So for example, if I have a datapoint it would either be (X, 1) or (X, 0).  I should be able to use one final unit, and the output should be close to 0 if the model is learning correctly and the ground truth is 0, etc.

**Comment by reesepathak** [comment]
what you're proposing is to convert my y vector's encoding from just 0 and 1 to: [1 0] if 0, and [0 1] is 1. I should not need to do that if I'm just doing binary classification though!

**Comment by aymericdamien** [comment]
yes! but that way, you will have 'accuracy' metric working. (You can see to_categorical function or regression 'to_categorical' argument).

**Comment by reesepathak** [comment]
alright, but is there no way to do so without having 2-vectors as my labels? I would prefer to have scalar labels to keep this similar to my Theano work in Keras.

**Comment by aymericdamien** [comment]
There is a way, you can use binary_crossentropy actually, but the 'accuracy' metric will have an issue. I will work on a fix to that.
